Objects/genericaliasobject.c

```diff
@@ -950,17 +973,117 @@ PyTypeObject Py_GenericAliasType = {
     .tp_vectorcall_offset = offsetof(gaobject, vectorcall),
 };
 
-PyObject *
-Py_GenericAlias(PyObject *origin, PyObject *args)
-{
+static PyObject *
+_Py_GenericAlias_impl_nocache(PyObject *origin, PyObject *args, bool starred) {
+    // args must be normalized at this point.
     gaobject *alias = (gaobject*) PyType_GenericAlloc(
             (PyTypeObject *)&Py_GenericAliasType, 0);
     if (alias == NULL) {
         return NULL;
     }
-    if (!setup_ga(alias, origin, args)) {
-        Py_DECREF(alias);
+    setup_ga(alias, origin, args, starred);
+    return (PyObject *)alias;
+}
+
+static PyObject *
+_Py_GenericAlias_impl(PyObject *origin, PyObject *args, bool starred) {
+    Py_hash_t hash;
+    bool unhashable = false;
+    PyObject *star = NULL;
+    PyObject *key = NULL;
+    PyObject *result = NULL;
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+
+    int arg_res = normalize_args(&args);
+    if (arg_res < 0) {
         return NULL;
     }
-    return (PyObject *)alias;
+
+    star = PyBool_FromLong(starred);
+    if (star == NULL) {
+        goto error;
+    }
+    key = PyTuple_Pack(3, origin, args, star);
+    if (key == NULL) {
+        goto error;
+    }
+
+    hash = PyObject_Hash(key);
+    if (hash == -1) {
+        // `key` contains unhashable objects, stop right here.
+        // Just return the object itself without setting cache.
+        PyErr_Clear();
+        unhashable = true;
+    } else {
+        result = _PyDict_GetItem_KnownHash(interp->genericalias_cache,
+                                            key, hash);
+        if (result) {
+            Py_INCREF(result);
+            goto finally;
+        }
+        if (PyErr_Occurred()) {
+            goto error;
+        }
+    }
```

Comment:
`args` are not released in these paths. L1022, L1025

Reply:
I wrote an extension to track down all ref count changes. Then I got here.

I am considering opening an issue to add this util into the main.